Utils/src/L1NonPrefiringProbProducer.cc

```diff
@@ -0,0 +1,348 @@
+
+#include "DataFormats/Math/interface/deltaR.h"
+#include "DataFormats/PatCandidates/interface/Jet.h"
+#include "DataFormats/PatCandidates/interface/Photon.h"
+#include "DataFormats/PatCandidates/interface/Muon.h"
+#include "TFile.h"
+#include "TH2.h"
+#include "TF1.h"
+
+#include "TreeMaker/Utils/interface/EnergyFractionCalculator.h"
+
+#include <iostream>
+#include <string>
+#include <vector>
+#include <array>
+#include <cmath>
+
+//based on: https://github.com/lathomas/cmssw/blob/L1Prefiring_9_4_9/L1Prefiring/EventWeightProducer/plugins/L1NonPrefiringProbProducer.cc
+
+namespace {
+   void dot(std::array<double,3>& a1, const std::array<double,3>& a2){
+      for(unsigned i = 0; i < a1.size(); ++i){
+         a1[i] *= a2[i];
+      }
+   }
+}
+
+class L1NonPrefiringProbProducer : public edm::global::EDProducer<> {
+public:
+  explicit L1NonPrefiringProbProducer(const edm::ParameterSet&);
+  
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+  
+private:
+  void produce(edm::StreamID, edm::Event&, const edm::EventSetup&) const override;
+  std::array<double,3> getNonPrefiringRateECAL( double eta, double pt, TH2F * h_prefmap) const;
+  std::array<double,3> getNonPrefiringRateMuon( double eta, double phi, double pt) const;
+ 
+  edm::EDGetTokenT<std::vector< pat::Photon> > photons_token_; 
+  edm::EDGetTokenT<std::vector< pat::Jet> > jets_token_;
+  edm::EDGetTokenT<std::vector< pat::Muon> > muons_token_;
+
+  TH2F* h_prefmap_photon; 
+  TH2F* h_prefmap_jet;
+  TF1* h_prefparam_muon_0p00to0p20;
+  TF1* h_prefparam_muon_0p20to0p30;
+  TF1* h_prefparam_muon_0p30to0p55;
+  TF1* h_prefparam_muon_0p55to0p83;
+  TF1* h_prefparam_muon_0p83to1p24;
+  TF1* h_prefparam_muon_1p24to1p40;
+  TF1* h_prefparam_muon_1p40to1p60;
+  TF1* h_prefparam_muon_1p60to1p80;
+  TF1* h_prefparam_muon_1p80to2p10;
+  TF1* h_prefparam_muon_2p10to2p25;
+  TF1* h_prefparam_muon_2p25to2p40;
+  TF1* h_prefparam_muon_hotspot;
+  std::string dataeraecal_;
+  std::string dataeramuon_;
+  bool useEMpt_;
+  double prefiringRateSystUncECAL_;
+  double prefiringRateSystUncMuon_;
+  double jet_max_muon_fraction_;
+
+};
+
+L1NonPrefiringProbProducer::L1NonPrefiringProbProducer(const edm::ParameterSet& iConfig) :
+  photons_token_(consumes<std::vector<pat::Photon>>(iConfig.getParameter<edm::InputTag>("ThePhotons"))),
+  jets_token_(consumes<std::vector<pat::Jet>>(iConfig.getParameter<edm::InputTag>("TheJets"))),
+  muons_token_(consumes<std::vector<pat::Muon>>(iConfig.getParameter<edm::InputTag>("TheMuons"))),
+  dataeraecal_(iConfig.getParameter<std::string>("DataEraECAL")),
+  dataeramuon_(iConfig.getParameter<std::string>("DataEraMuon")),
+  useEMpt_(iConfig.getParameter<bool>("UseJetEMPt")),
+  prefiringRateSystUncECAL_(iConfig.getParameter<double>("PrefiringRateSystematicUnctyECAL")),
+  prefiringRateSystUncMuon_(iConfig.getParameter<double>("PrefiringRateSystematicUnctyMuon")),
+  jet_max_muon_fraction_(iConfig.getParameter<double>("JetMaxMuonFraction"))
```

Comment:
this is not used. should it be used somewhere? if not, delete it.

Reply:
Thanks for the check, it should be used, but I forgot to implement, will appear in next commit.